### PR TITLE
Added support for running services fully backgrounded on macOS

### PR DESCRIPTION
**Pull request by GottemHams**
Re: #485

After the changes from this PR I first reverted the plist to the old state, so that it doesn't contain the `LimitLoadToSessionType` key. Then I tried to start Apache (remember, `brew` is my alias that wraps `sudo su`):
```
myuser:~$ grep LimitLoadToSessionType /opt/homebrew/opt/httpd/homebrew.mxcl.httpd.plist
myuser:~$ brew services start httpd
Could not enable service: 125: Domain does not support specified action
Error: Failure while executing; `/bin/launchctl enable gui/503/homebrew.mxcl.httpd` exited with 125.
```

This is exactly the old/original behaviour (note the `gui` domain). Just to be sure I physically logged on to `mybrew`, then:
```
mybrew:~$ grep LimitLoadToSessionType /opt/homebrew/opt/httpd/homebrew.mxcl.httpd.plist
mybrew:~$ brew services start httpd
==> Successfully started `httpd` (label: homebrew.mxcl.httpd)
```

After logging out the service gets killed, just what you'd expect since we're still on the `gui` domain.

Back to my own user:
```
myuser:~$ brew reinstall httpd
..............................
myuser:~$ grep LimitLoadToSessionType /opt/homebrew/opt/httpd/homebrew.mxcl.httpd.plist
	<key>LimitLoadToSessionType</key>
myuser:~$ brew services start httpd
==> Successfully started `httpd` (label: homebrew.mxcl.httpd)
```

---

Some notes:
1. I'm not quite sure what to do about adding a test for the old plist behaviour, since that would require a (pre-installed) plist that doesn't have the `LimitLoadToSessionType` key (yet). I could add something fixed but I don't think that's the best way.

2. I don't really like the name `domain_target_needs_background?` for the function I added but I couldn't think of something more fitting either. I could just add the code directly to `def domain_target` instead, but I wanted to keep that as minimal as possible. Also, if we're gonna keep the second function did you want it documented in the `system_spec.rb`? It's sort of an internal function though.

3. For the `domain_target` spec you had an explicit capturing group for the UID, so for distinguishing `gui` and `user` I used a non-capturing group just to be sure it doesn't interfere with anything expecting the first group to be the UID.

---

Seems like the spec tests are failing, but looking at the spec for `user_of_process` I don't see anything to specify the arguments?

### CODE REVIEW

File: spec/homebrew/system_spec.rb [hunk 1]
```diff
@@ -75,7 +75,7 @@
   describe "#domain_target" do
     it "returns the current domain target" do
       allow(described_class).to receive(:root?).and_return(false)
-      expect(described_class.domain_target).to match(%r{gui/(\d+)})
+      expect(described_class.domain_target).to match(%r{(?:gui|user)/(\d+)})
```
Comment (MikeMcQuaid) [inline spec/homebrew/system_spec.rb:78]: Would be good to expand this test to catch the three differing values.

File: lib/service/system.rb [hunk 2]
```diff
@@ -72,9 +72,21 @@ def path
       root? ? boot_path : user_path
     end
 
-    def domain_target
+    def domain_target_needs_background?(service)
+      plist_data = service.generate_plist(nil)
+      plist = begin
+        Plist.parse_xml(plist_data)
+      rescue
```
Comment (MikeMcQuaid) [inline lib/service/system.rb:79]: A comment here explaining why it's `nil` would be nice.

File: lib/service/system.rb [hunk 3]
```diff
@@ -72,9 +72,21 @@ def path
...
+        nil
+      end
+      !plist.nil? && !plist["LimitLoadToSessionType"].nil?
```
Comment (MikeMcQuaid) [inline lib/service/system.rb:82]:
```suggestion
      plist.present? && plist["LimitLoadToSessionType"].present?
```

**Review by MikeMcQuaid — COMMENTED** [review]
Makes sense to me so far. Just so I understand: this will basically only kick in if the plist supports it, right?

**Comment by MikeMcQuaid** [comment]
> I'm not quite sure what to do about adding a test for the old plist behaviour, since that would require a (pre-installed) plist that doesn't have the LimitLoadToSessionType key (yet). I could add something fixed but I don't think that's the best way.

Stubbing/mocking the plist values seems like it'd make sense here 👍🏻 

> I don't really like the name domain_target_needs_background? for the function I added but I couldn't think of something more fitting either.

Seems OK to me!

> For the domain_target spec you had an explicit capturing group for the UID, so for distinguishing gui and user I used a non-capturing group just to be sure it doesn't interfere with anything expecting the first group to be the UID.

Think it'd be good to separate this into multiple test for each branch.

**Comment by GottemHams** [comment]
I moved the `user` domain test to its own branch and rearranged the 3 tests so it reflects the actual code. Took a bit to properly figure out the test syntax though. ;_;

Also yes, if the plist has not been rewritten it will (continue to) use the default `gui` domain. As for testing the old plist behaviour: I think it made the most sense to have that handled by the GitHub CI runner. The spec tests would otherwise simply only check the domain target being `gui` and not if it can actually be started.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 name: CI
 on:
   push:
-    branches: master
+    branches: [master]
```
Comment (SMillerDev) [inline .github/workflows/tests.yml:4]: Why is this needed?

File: lib/service/system.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply lib/service/system.rb:81]: This is not resolved yet.

**Review by MikeMcQuaid — COMMENTED** [review]
Looking good so far! A few minor tweaks needed.

File: .github/workflows/tests.yml (hunk 4 above)
Comment (GottemHams) [reply .github/workflows/tests.yml:4]: It's not really, I had to add my `test` branch so I could check the CI without affecting this PR. Obviously the branch is now removed from the YAML but I left the array markers. I can remove those if you want, but either way the YAML is valid.

File: lib/service/system.rb (hunk 2 above)
Comment (GottemHams) [reply lib/service/system.rb:81]: I did add [a comment explaining that](https://github.com/Homebrew/homebrew-services/pull/512/files#diff-628d1358d1f6da549a3320bbf39688c4f7800fb6e7c0143d252549c7f39c928f), I think it's just barely not shown here because you tied your comment to line `plist_data = ...` line and not `def domain_target_needs_background?(service)`. :>

Unless you meant why `nil` under `rescue`? In that case: isn't it already pretty clear that it indicates a parsing error and we shouldn't/can't even check the particular key?

Comment (MikeMcQuaid) [reply lib/service/system.rb:81]: > Unless you meant why nil under rescue? In that case: isn't it already pretty clear that it indicates a parsing error and we shouldn't/can't even check the particular key?

Yes, a comment directly above the `nil` would be preferred. Ideally, also catch a more specific exception.

Comment (GottemHams) [reply lib/service/system.rb]: Alright, I think that should about cover it.

**Review by SMillerDev — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]
Nice work, thanks again @GottemHams!
